numpy/lib/function_base.py

```diff
@@ -4912,6 +4912,21 @@ def trapz(y, x=None, dx=1.0, axis=-1):
     return ret
 
 
+if overrides.ARRAY_FUNCTION_ENABLED:
+    # Below wrapping ensures `__code__` attribute;  SciPy <=1.10 expected a
+    # python func (which have it): https://github.com/scipy/scipy/issues/17811
+    assert not hasattr(trapz, "__code__")
+
+    _trapz_orig = trapz
+
+    @functools.wraps(_trapz_orig)
+    def trapz(y, x=None, dx=1.0, axis=-1):
+        # Hack so that SciPy has some python code it can steal
+        return _trapz_orig(y, x=x, dx=dx, axis=axis)
```

Comment:
Ah, this means we always pass defaults in `__array_function__` (which is probably OK).  We can also go back to `*args, **kwargs`.  Or, I can just override all the attributes used by SciPy (`__code__`, and a couple of others).

Reply:
Finding @dhomeier [comment at astropy](https://github.com/astropy/astropy/issues/14287#issuecomment-1397729552):

> ```python
> E           astropy.utils.exceptions.AstropyWarning: function 'trapz' is not known to astropy's Quantity. Will run it anyway, hoping it will treat ndarray subclasses correctly. Please raise an issue at https://github.com/astropy/astropy/issues.
> ```
> 
> The issue seems to be that with the proposed fix np.trapz identifies as _trapz_orig, which is not in function_helpers.SUBCLASS_SAFE_FUNCTIONS.

Wrapping things up seems to confuse astropy a bit (the function is just an entry-point).

So I will fix this up to just steal all the attributes so that pretending that this is a Python function works (what SciPy does) to allow creating a new Python function from the non-python version.

Reply:
Thanks @seberg, this version resolves astropy/astropy#14287 out of the box!